File: 054_MinimizeMaxDistanceToGasStation.cpp
// Topic: Binary Search / Greedy
// Minimize Max Distance to Gas Station: binary search on distance and check required stations.
// Time Complexity: O(n log R) where R is range precision
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int requiredStations(const vector<int>& stations, double D)
{
    int cnt = 0;
    for (int i = 1; i < stations.size(); ++i)
    {
        double dist = stations[i] - stations[i - 1];
        cnt += (int)(dist / D) - 1;  // careful with floor
    }
    return cnt;
}

double minmaxGasDist(vector<int>& stations, int k)
{
    double lo = 0, hi = stations.back() - stations.front();
    for (int it = 0; it < 60; ++it)
    {
        double mid = (lo + hi) / 2;
        int need = 0;
        for (int i = 1; i < stations.size(); ++i)
            need += (int)((stations[i] - stations[i - 1]) / mid) - 1;
        if (need > k)
            lo = mid;
        else
            hi = mid;
    }
    return hi;
}

int main()
{
    cout << fixed << setprecision(6)
         << minmaxGasDist((vector<int>{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}), 9) << "\n";
    cout << fixed << setprecision(6) << minmaxGasDist((vector<int>{0, 10}), 9) << "\n";
    return 0;
}

File: 054_MinimizeMaxDistanceToGasStation.py
"""
Topic: Binary Search / Greedy
Problem: Minimize Max Distance to Gas Station
Description: Binary search the maximum allowed distance and check feasibility.
Time: O(n log W) where W is search range precision
Space: O(1)
"""

from typing import List


def minmaxGasDist(stations: List[int], k: int) -> float:
    def feasible(d: float) -> bool:
        # count how many extra stations needed for gap with max distance d
        need = 0
        for i in range(1, len(stations)):
            gap = stations[i] - stations[i-1]
            need += int(gap // d)  # number of pieces larger than d
            if need > k:
                return False
        return need <= k

    lo, hi = 1e-6, stations[-1] - stations[0]
    for _ in range(50):
        mid = (lo + hi) / 2
        if feasible(mid):
            hi = mid
        else:
            lo = mid
    return hi


if __name__ == '__main__':
    print(round(minmaxGasDist([1,2,3,4,5,6,7,8,9,10], 9), 6))
    print(round(minmaxGasDist([23, 32, 40, 47, 79, 84, 100], 1), 6))
